### PR TITLE
Samples for CustomMapping, IndicateMissingValues, ReplaceMissingValues

### LINKED ISSUE #3117: Need a sample for saving and loading CustomMapping Transform
We currently only have a simple sample to use the custom mapping transform, but saving and loading is a bit more involved and a sample about that would be very useful.

Here is the current sample that we have:
https://github.com/dotnet/machinelearning/blob/master/docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSample.cs

Here is a test that shows how to save and load the custom mapping transform:
https://github.com/dotnet/machinelearning/blob/366332047340bdb12cf5fe2a11f9a6e257af9b84/test/Microsoft.ML.Tests/Transformers/CustomMappingTests.cs#L33-L85

### PR DESCRIPTION
**Pull request by artidoro**
Related to #1209 
Fixes #3117 

Made samples for the multi-column setting of:

- `ReplaceMissingValues`
- `IndicateMissingValues`

I also made a sample to save and load the `CustomMapping` estimator.

### CODE REVIEW

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSample.cs [hunk 1]
```diff
@@ -1,4 +1,7 @@
 ﻿using System;
+using System.Collections.Generic;
+using Microsoft.ML.Transforms;
+
 namespace Microsoft.ML.Samples.Dynamic
```
Comment (abgoswam) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSample.cs]: >Microsoft.ML [](start = 10, length = 12)

we moved to a pattern of adding a "using Microsoft.ML"  , instead of adding  it in the namespace

**Comment by shmoradims** [comment]
> using System;

please remove Sample from filenames: CusomtMapingSample -> CustomMapping

Same with the other file

---
Refers to: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSample.cs:1 in d32f027. [](commit_id = d32f02787e9a6d9dbaf9826c49dfe9e24250db9b, deletion_comment = False)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSample.cs (hunk 1 above)
Comment (shmoradims) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSample.cs]: see [https://github.com/dotnet/machinelearning/issues/3205](https://github.com/dotnet/machinelearning/issues/3205)

---
In reply to: [272664635](https://github.com/dotnet/machinelearning/pull/3216#discussion_r272664635) [](ancestors = 272664635)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSample.cs [hunk 2]
```diff
@@ -10,71 +13,88 @@ public static void Example()
             var mlContext = new MLContext();
 
             // Get a small dataset as an IEnumerable and convert it to an IDataView.
-            var data = SamplesUtils.DatasetUtils.GetInfertData();
-            var trainData = mlContext.Data.LoadFromEnumerable(data);
+            var rawData = GetData();
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSample.cs]: >GetData [](start = 26, length = 7)

if the dataset is has <10 rows, please define it inline instead of a method.

we only use a method for generating 100+ random data point (usually for trainer.)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSample.cs [hunk 3]
```diff
@@ -10,71 +13,88 @@ public static void Example()
...
+            // Printing the input data. 
+            Console.WriteLine("Age\t Salary");
+            foreach (var row in rawData)
+                Console.WriteLine($"{row.Age}\t {row.Salary}");
+            // Expected output:
+            // Age      Salary
+            // 26       40000
+            // 35       80000
+            // 34       10000
+            // 28       100000
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSample.cs]: if you define the data inline, you don't need this preview. it's already obvious from code.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSample.cs [hunk 4]
```diff
@@ -10,71 +13,88 @@ public static void Example()
...
+            Action<InputData, CustomMappingOutput > mapping =
                 (input, output) => output.IsUnderThirty = input.Age < 30;
 
-            // Custom transformations can be used to transform data directly, or as part of a pipeline. Below we transform data directly.
-            var estimator = mlContext.Transforms.CustomMapping(mapping, null);
-            var transformedData = estimator.Fit(trainData).Transform(trainData);
+            // Custom transformations can be used to transform data directly, or as part of a pipeline of estimators.
+            // Note: If contractName is null in the CustomMapping estimator, any pipeline of estimators containing it, cannot be saved and loaded back. 
+            // See other sample on how to load and save the CustomMapping estimator.
+            var estimator = mlContext.Transforms.CustomMapping(mapping, contractName: null);
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSample.cs]: >estimator [](start = 16, length = 9)

we call these 'pipeline' in the rest of sample, eventhough the only have a single estimator

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSampleSaveAndLoad.cs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+            // Custom transformations can be used to transform data directly, or as part of a pipeline. Below we transform data directly.
+            var estimator = mlContext.Transforms.CustomMapping(new IsUnderThirtyCustomAction().GetMapping(), contractName: "IsUnderThirty");
+            var transform = estimator.Fit(data);
+
+            // To save and load the CustomMapping estimator, the assembly in which the custom action is defined needs to be registered in the
+            // environment. The following registers the assembly where IsUnderThirtyCustomAction is defined.    
+            mlContext.ComponentCatalog.RegisterAssembly(typeof(IsUnderThirtyCustomAction).Assembly);
+
+            // Now the transform pipeline can be saved and loaded through the usual MLCOntext method.
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSampleSaveAndLoad.cs]: >MLCOntext [](start = 84, length = 9)

case typo

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValuesMultiColumn.cs [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+                    $" MissingReplaced1: {vectorPrinter(row.MissingReplaced1.Cast<object>().ToArray())}" +
+                    $" MissingReplaced2: {vectorPrinter(row.MissingReplaced2.Cast<object>().ToArray())}");
+            }
+
+            // Expected output:
+            // Notice how the NaN of the Features column for the second row is replaced by the mean of (1, 2, 6) the values in that row
+            // 
+            // Label: 3 Features1: [1 1 0] Features2: [1 1]  MissingReplaced1: [1 1 0] MissingReplaced2: [1 1]
+            // Label: 32 Features1: [0 NaN 1] Features2: [0 1]  MissingReplaced1: [0 3 1] MissingReplaced2: [0 1]
+            // Label: 5 Features1: [-1 2 - 3] Features2: [-1 NaN]  MissingReplaced1: [-1 2 - 3] MissingReplaced2: [-1 1]
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValuesMultiColumn.cs]: >- 3 [](start = 41, length = 3)

VS has added some extra spaces during pasting

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValuesMultiColumn.cs [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+                    $" MissingReplaced1: {vectorPrinter(row.MissingReplaced1.Cast<object>().ToArray())}" +
+                    $"MissingReplaced2: {vectorPrinter(row.MissingReplaced2.Cast<object>().ToArray())}");
+            }
+
+            // Expected output:
+            // Notice how the NaN of the Features column for the second row is replaced by 0, the default value for floats.
+            // 
+            // Label: 3 Features1: [1 1 0] Features2: [1 1]  MissingReplaced1: [1 1 0]MissingReplaced2: [1 1]
+            // Label: 32 Features1: [0 NaN 1] Features2: [0 1]  MissingReplaced1: [0 0 1]MissingReplaced2: [0 1]
+            // Label: 5 Features1: [-1 2 - 3] Features2: [-1 NaN]  MissingReplaced1: [-1 2 - 3]MissingReplaced2: [-1 0]
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValuesMultiColumn.cs]: >- 3 [](start = 41, length = 3)

ditto: - 3 -> -3 and in the rest of these 4 lines of comment

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValuesMultiColumn.cs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+            var meanTransformer = meanPipeline.Fit(data);
+            var meanTransformedData = meanTransformer.Transform(data);
+
+            // We can extract the newly created column as an IEnumerable of SampleDataTransformed, the class we define below.
+            var meanRowEnumerable = mlContext.Data.CreateEnumerable<SampleDataTransformed>(meanTransformedData, reuseRowObject: false);
+
+            // ReplaceMissingValues is used to create a column where missing values are replaced according to the ReplacementMode.
+            var defaultPipeline = mlContext.Transforms.ReplaceMissingValues(new[] {
+                new InputOutputColumnPair("MissingReplaced1", "Features1"),
+                new InputOutputColumnPair("MissingReplaced2", "Features2")
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValuesMultiColumn.cs]: optional: we don't have an in-place transformation multicolumn sample. can we transform the data in-place and not have MissingReplaced1/MissingReplaced2?

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValuesMultiColumn.cs [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+            // a small printing utility
+            Func<object[], string> vectorPrinter = (object[] vector) =>
+            {
+                string preview = "[";
+                foreach (var slot in vector)
+                    preview += $"{slot} ";
+                return preview += "]";
+
+            };
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValuesMultiColumn.cs]: this one can go to the bottom of the file. all the helper utils are secondary, and should go below Example()

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValuesMultiColumn.cs [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+            // And finally, we can write out the rows of the dataset, looking at the columns of interest.
+            foreach (var row in meanRowEnumerable)
+            {
+                Console.WriteLine($"Label: {row.Label} Features1: {vectorPrinter(row.Features1.Cast<object>().ToArray())}" +
+                    $" Features2: {vectorPrinter(row.Features2.Cast<object>().ToArray())} " +
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValuesMultiColumn.cs]: >Cast<object> [](start = 63, length = 12)

this Cast is only required for printing, but it's making the sample look complicated. please push it to the helper print function. also the print function can take float[], can't it?

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValuesMultiColumn.cs [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+                string preview = "[";
+                foreach (var slot in vector)
+                    preview += $"{slot} ";
+                return preview += "]";
+
+            };
+
+            // And finally, we can write out the rows of the dataset, looking at the columns of interest.
+            foreach (var row in meanRowEnumerable)
+            {
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValuesMultiColumn.cs]: >{ [](start = 12, length = 1)

omit for single-line blocks

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValuesMultiColumn.cs [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+            // 
+            // Label: 3 Features1: [1 1 0] Features2: [1 1]  MissingReplaced1: [1 1 0]MissingReplaced2: [1 1]
+            // Label: 32 Features1: [0 NaN 1] Features2: [0 1]  MissingReplaced1: [0 0 1]MissingReplaced2: [0 1]
+            // Label: 5 Features1: [-1 2 - 3] Features2: [-1 NaN]  MissingReplaced1: [-1 2 - 3]MissingReplaced2: [-1 0]
+            // Label: 9 Features1: [-1 6 - 3] Features2: [0 ∞ ]  MissingReplaced1: [-1 6 - 3]MissingReplaced2: [0 ∞ ]
+        }
+
+        private class DataPoint
+        {
+            public float Label { get; set; }
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValuesMultiColumn.cs]: please drop fields that aren't used in the sample

File: src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs [hunk 13]
```diff
@@ -1020,11 +1020,12 @@ private void LoadExtensions(Assembly assembly, bool throwOnError)
         /// Gets a value indicating whether <paramref name="assembly"/> can contain extensions.
         /// </summary>
         /// <remarks>
-        /// All ML.NET product assemblies won't contain extensions.
+        /// All ML.NET product assemblies won't contain extensions besides Microsoft.ML.Samples.
         /// </remarks>
         private static bool CanContainExtensions(Assembly assembly)
         {
             if (assembly.FullName.StartsWith("Microsoft.ML.", StringComparison.Ordinal)
+                && !assembly.FullName.StartsWith("Microsoft.ML.Samples", StringComparison.Ordinal)
```
Comment (shmoradims) [inline src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs]: >   && !assembly.FullName.StartsWith("Microsoft.ML.Samples", StringComparison.Ordinal) [](start = 13, length = 85)

code change detected!!! could you do this as a separate PR with proper shiproom triage? we don't want to mix code-change with documentation PRs. :)

File: src/Microsoft.ML.Transforms/CustomMappingCatalog.cs [hunk 14]
```diff
@@ -27,6 +27,7 @@ public static class CustomMappingCatalog
         /// <format type="text/markdown">
         /// <![CDATA[
         ///  [!code-csharp[FastTree](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMapping.cs)]
+        ///  [!code-csharp[FastTree](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSampleSaveAndLoad.cs)]
```
Comment (shmoradims) [inline src/Microsoft.ML.Transforms/CustomMappingCatalog.cs]: >FastTree [](start = 27, length = 8)

incorrect

Comment (shmoradims) [inline src/Microsoft.ML.Transforms/CustomMappingCatalog.cs]: >Sample [](start = 121, length = 6)

please make sure you update this, when your remove 'Sample' from filename.

File: src/Microsoft.ML.Transforms/ExtensionsCatalog.cs [hunk 15]
```diff
@@ -38,6 +38,12 @@ public static class ExtensionsCatalog
         /// </summary>
         /// <param name="catalog">The transform extensions' catalog.</param>
         /// <param name="columns">Specifies the names of the columns on which to apply the transformation.</param>
+        /// <example>
+        /// <format type="text/markdown">
+        /// <![CDATA[
+        ///  [!code-csharp[RPCA](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/IndicateMissingValuesMultiColumn.cs)]
```
Comment (shmoradims) [inline src/Microsoft.ML.Transforms/ExtensionsCatalog.cs]: >RPCA [](start = 27, length = 4)

incorrect

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSample.cs [hunk 16]
```diff
@@ -10,71 +13,88 @@ public static void Example()
...
 
-            // Here instead we use it as part of a pipeline of estimators.
-            var pipeline = mlContext.Transforms.CustomMapping(mapping, null)
-                .Append(mlContext.Transforms.Concatenate(outputColumnName: "Features", inputColumnNames: new[] { "Parity", "Induced" }))
-                // It is useful to add a caching checkpoint before a trainer that does several passes over the data.
-                .AppendCacheCheckpoint(mlContext)
-                // We use binary FastTree to predict the label column that was generated by the custom mapping at the first step of the pipeline.
-                .Append(mlContext.BinaryClassification.Trainers.FastTree(labelColumnName: "IsUnderThirty"));
+        // Defines only the column to be generated by the custom mapping transformation in addition to the columns already present.
+        public class CustomMappingOutput
```
Comment (wschin) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSample.cs]:
```suggestion
        private class CustomMappingOutput
```

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSample.cs [hunk 17]
```diff
@@ -10,71 +13,88 @@ public static void Example()
...
+        {
+            public bool IsUnderThirty { get; set; }
+        }
 
-            // We can train the pipeline and use it to transform data.
-            transformedData = pipeline.Fit(trainData).Transform(trainData);
+        // Defines the schema of the input data.
+        public class InputData
```
Comment (wschin) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSample.cs]:
```suggestion
        private class InputData
```

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSample.cs [hunk 18]
```diff
@@ -10,71 +13,88 @@ public static void Example()
...
+        {
+            public float Age { get; set; }
+            public float Salary { get; set; }
         }
 
-        // This defines only the column to be generated by the transformation in addition to the columns already present.
-        public class OutputRow
+        // Defines the schema of the transformed data, which includes the new column IsUnderThirty.
+        public class TransformedData
```
Comment (wschin) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSample.cs]: May we change all example classes to private?
```suggestion
        private class TransformedData
```

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSample.cs (hunk 1 above)
Comment (artidoro) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSample.cs]: Thank you!


---
In reply to: [272759050](https://github.com/dotnet/machinelearning/pull/3216#discussion_r272759050) [](ancestors = 272759050,272664635)

**Comment by artidoro** [comment]
Thank you for the review comments, I have update the code accordingly.
I had to change the signature of the samples assembly, otherwise it was not possible to define a custom map and save and load the custom estimator that uses it.

File: docs/samples/Microsoft.ML.Samples/Microsoft.ML.Samples.csproj [hunk 19]
```diff
@@ -1,8 +1,10 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
-
+  
   <PropertyGroup>
     <TargetFramework>netcoreapp2.1</TargetFramework>
     <OutputType>Exe</OutputType>
+    <AssemblyOriginatorKeyFile>$(ToolsDir)Test.snk</AssemblyOriginatorKeyFile>
```
Comment (artidoro) [inline docs/samples/Microsoft.ML.Samples/Microsoft.ML.Samples.csproj]: By default the assemblies are signed using Open.snk. However, if we want to register an assembly in the component catalog using `mlContext.ComponentCatalog.RegisterAssembly()`, the assembly needs to pass the following condition:
https://github.com/dotnet/machinelearning/blob/b861b5d64841cbe0f2c866ee7586872aac450a51/src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs#L1025-L1034

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMapping.cs [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+                new InputData { Age = 28 },
+            };
+            var data = mlContext.Data.LoadFromEnumerable(samples);
+
+            // We define the custom mapping between input and output rows that will be applied by the transformation.
+            Action<InputData, CustomMappingOutput > mapping =
+                (input, output) => output.IsUnderThirty = input.Age < 30;
+
+            // Custom transformations can be used to transform data directly, or as part of a pipeline of estimators.
+            // Note: If contractName is null in the CustomMapping estimator, any pipeline of estimators containing it, cannot be saved and loaded back.
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMapping.cs]: >annot be [](start = 120, length = 8)

break

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSaveAndLoad.cs [hunk 21]
```diff
@@ -0,0 +1,88 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.ML;
+using Microsoft.ML.Transforms;
+
+namespace Samples.Dynamic
+{
+    public static class CustomMappingSaveAndLoad
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSaveAndLoad.cs]: >CustomMappingSaveAndLoad [](start = 24, length = 24)

Do we really need two samples? They are identical except for the serialization / deserialization stage.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMapping.cs [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+            };
+            var data = mlContext.Data.LoadFromEnumerable(samples);
+
+            // We define the custom mapping between input and output rows that will be applied by the transformation.
+            Action<InputData, CustomMappingOutput > mapping =
+                (input, output) => output.IsUnderThirty = input.Age < 30;
+
+            // Custom transformations can be used to transform data directly, or as part of a pipeline of estimators.
+            // Note: If contractName is null in the CustomMapping estimator, any pipeline of estimators containing it, cannot be saved and loaded back. 
+            // See other sample on how to load and save the CustomMapping estimator.
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMapping.cs]: >// See other sample on how to load and save the CustomMapping estimator. [](start = 12, length = 72)

I would avoid referencing other samples from samples because it's hard to maintain.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/IndicateMissingValues.cs [hunk 23]
```diff
@@ -1,32 +1,31 @@
...
             };
-            // Convert training data to IDataView, the general data type used in ML.NET.
             var data = mlContext.Data.LoadFromEnumerable(samples);
 
-            // IndicateMissingValues is used to create a boolean containing
-            // 'true' where the value in the input column is NaN. This value can be used
-            // to replace missing values with other values.
-            IEstimator<ITransformer> pipeline = mlContext.Transforms.IndicateMissingValues("MissingIndicator", "Features");
+            // IndicateMissingValues is used to create a boolean containing 'true' where the value in the 
+            // input column is missing. For floats, missing values are float.NaN.
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/IndicateMissingValues.cs]: > are  [](start = 66, length = 5)

are represented as

Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/IndicateMissingValues.cs]: >floats [](start = 44, length = 6)

floats and doubles => NaN (not float.NaN)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/IndicateMissingValues.cs [hunk 24]
```diff
@@ -1,32 +1,31 @@
...
+            // The newly created value can be used to replace missing values with other values.
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/IndicateMissingValues.cs]: >// The newly created value can be used to replace missing values with other values. [](start = 12, length = 83)

Indicator isn't used to replace. It's used to give extra information to the learner.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/IndicateMissingValuesMultiColumn.cs [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+            {
+                new DataPoint(){ Features1 = new float[3] {1, 1, 0}, Features2 = new float[2] {1, 1} },
+                new DataPoint(){ Features1 = new float[3] {0, float.NaN, 1}, Features2 = new float[2] {float.NaN, 1} },
+                new DataPoint(){ Features1 = new float[3] {-1, float.NaN, -3}, Features2 = new float[2] {1, float.PositiveInfinity} },
+            };
+            var data = mlContext.Data.LoadFromEnumerable(samples);
+
+            // IndicateMissingValues is used to create a boolean containing 'true' where the value in the 
+            // input column is missing. For floats, missing values are float.NaN.
+            // We can use an array of InputOutputColumnPair to apply the MissingValueIndicatorEstimator to multiple columns in one pass over the data.
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/IndicateMissingValuesMultiColumn.cs]: >ass over the data. [](start = 132, length = 18)

long

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValues.cs [hunk 26]
```diff
@@ -13,27 +13,18 @@ public static void Example()
...
-
-            // Now we can transform the data and look at the output to confirm the behavior of the estimator.
-            // This operation doesn't actually evaluate data until we read the data below.
-            var meanTransformer = meanPipeline.Fit(data);
-            var meanTransformedData = meanTransformer.Transform(data);
-
-            // We can extract the newly created column as an IEnumerable of SampleDataTransformed, the class we define below.
-            var meanRowEnumerable = mlContext.Data.CreateEnumerable<SampleDataTransformed>(meanTransformedData, reuseRowObject: false);
-
+            // Default ReplaceMode:
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValues.cs]: >// Default ReplaceMode: [](start = 12, length = 23)

Delete

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValues.cs [hunk 27]
```diff
@@ -13,27 +13,18 @@ public static void Example()
...
             // ReplaceMissingValues is used to create a column where missing values are replaced according to the ReplacementMode.
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValues.cs]: >// ReplaceMissingValues  [](start = 12, length = 24)

// Here we use the default replacement mode, which replaces the value with the default value for its `type`.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValues.cs [hunk 28]
```diff
@@ -45,49 +36,41 @@ public static void Example()
...
-            //Label: 3  Features: [1 1    0] MissingReplaced: [1  1  0]
-            //Label: 32 Features: [0 NaN  1] MissingReplaced: [0  3  1]
-            //Label: 5  Features: [-1 2 - 3] MissingReplaced: [-1 2 -3]
-            //Label: 9  Features: [-1 6 - 3] MissingReplaced: [-1 6 -3]
+            // Features: [∞, 1, 0]      MissingReplaced: [∞, 1, 0]
+            // Features: [0, NaN, 1]    MissingReplaced: [0, 0, 1]
+            // Features: [-1, 2, -3]    MissingReplaced: [-1, 2, -3]
+            // Features: [-1, NaN, -3]  MissingReplaced: [-1, 0, -3]
+
+            // Mean ReplaceMode:
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValues.cs]: >// Mean ReplaceMode: [](start = 12, length = 20)

Isn't there one more mode?

File: docs/samples/Microsoft.ML.Samples/Microsoft.ML.Samples.csproj (hunk 19 above)
Comment (artidoro) [reply docs/samples/Microsoft.ML.Samples/Microsoft.ML.Samples.csproj]: After discussing with Eric, it could be even simpler not to sign this assembly. We don't need strong naming for the Samples assembly as it should not be referenced by external code anyways.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValues.cs (hunk 28 above)
Comment (artidoro) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValues.cs]: There is also maximum and minimum, but I don't think it adds much to add them to this sample.

---
In reply to: [273727261](https://github.com/dotnet/machinelearning/pull/3216#discussion_r273727261) [](ancestors = 273727261)

Comment (rogancarr) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValues.cs]: For the documentation, if we have more than one, we should have all of them &mdash; right now, it looks inconsistent or incomplete.

---
In reply to: [273735371](https://github.com/dotnet/machinelearning/pull/3216#discussion_r273735371) [](ancestors = 273735371,273727261)

Comment (artidoro) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ReplaceMissingValues.cs]: I would rather remove the mean replacement mode in this sample then.
We are not trying to exhaust all the possible settings in our samples. Would that be fine?

---
In reply to: [273736506](https://github.com/dotnet/machinelearning/pull/3216#discussion_r273736506) [](ancestors = 273736506,273735371,273727261)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSaveAndLoad.cs [hunk 29]
```diff
@@ -0,0 +1,88 @@
...
+        private class InputData
+        {
+            public float Age { get; set; }
+        }
+
+        // Defines the schema of the transformed data, which includes the new column IsUnderThirty.
+        private class TransformedData : InputData
+        {
+            public bool IsUnderThirty { get; set; }
+
```
Comment (wschin) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSaveAndLoad.cs]: Extra empty line.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSaveAndLoad.cs (hunk 21 above)
Comment (wschin) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSaveAndLoad.cs]:
```suggestion
    // This sample shows how to create a CustomMappingTransform, save it, load it, and finally run it.
    public static class CustomMappingSaveAndLoad
```

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/IndicateMissingValuesMultiColumn.cs [hunk 30]
```diff
@@ -0,0 +1,68 @@
...
+
+            // And finally, we can write out the rows of the dataset, looking at the columns of interest.
+            foreach (var row in rowEnumerable)
+                Console.WriteLine($"Features1: [{string.Join(", ", row.Features1)}]\t MissingIndicator1: [{string.Join(", ", row.MissingIndicator1)}]\t " +
+                    $"Features2: [{string.Join(", ", row.Features2)}]\t MissingIndicator2: [{string.Join(", ", row.MissingIndicator2)}]");
+
+            // Expected output:
+            // Features1: [1, 1, 0]     MissingIndicator1: [False, False, False]        Features2: [1, 1]       MissingIndicator2: [False, False]
+            // Features1: [0, NaN, 1]   MissingIndicator1: [False, True, False]         Features2: [NaN, 1]     MissingIndicator2: [True, False]
+            // Features1: [-1, NaN, -3]         MissingIndicator1: [False, True, False]         Features2: [1, ∞]       MissingIndicator2: [False, False]
```
Comment (singlis) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/IndicateMissingValuesMultiColumn.cs]: NIT - spacing for MissingIndicator doesnt align with the above lines.
